pkg/apis/azure/validation/secrets.go

```diff
@@ -33,6 +36,20 @@ func ValidateCloudProviderSecret(secret *corev1.Secret) error {
 		if len(val) == 0 {
 			return fmt.Errorf("field %q in secret %s cannot be empty", key, secretKey)
 		}
+		switch key {
+		case azure.SubscriptionIDKey, azure.TenantIDKey, azure.ClientIDKey:
+			// subscriptionID, tenantID, and clientID must be valid UUIDs
```

Comment:
Could you point the source of this information?

Reply:
At least for subsciption ID, I found that several Azure REST APIs reveal the following regex - `^[0-9A-Fa-f]{8}-([0-9A-Fa-f]{4}-){3}[0-9A-Fa-f]{12}$`. See https://docs.microsoft.com/en-us/rest/api/securitycenter/locations/get, https://docs.microsoft.com/en-us/rest/api/securitycenter/tasks/get-resource-group-level-task and few others Azure REST API docs.
I assume that this regex fits with the validation performed by `github.com/google/uuid`.

Reply:
I could not find authoritative documentation (from Microsoft) stating this explicitly. What I could find are various documents stating that these IDs come from "Azure Active Directory", as well as references in documents (both from Microsoft and 3rd parties) referring to them as "GUIDs", see e.g. 
* https://docs.microsoft.com/en-us/powershell/module/azurerm.resources/get-azurermadapplication?view=azurermps-6.13.0
* https://stackoverflow.com/questions/56551171/is-azure-aad-application-id-unique-in-whole-azure
* https://docs.bmc.com/docs/cloudlifecyclemanagement/46/setting-up-a-tenant-id-client-id-and-client-secret-for-azure-resource-manager-provisioning-669202145.html#:~:text=The%20subscription%20ID%20is%20a,along%20with%20the%20subscription%20ID.

We also know from looking at actual credentials that they are always GUIDs.

I know this is a bit weak. If we want to be on the safe side, we should perhaps refrain from validating as GUIDs and only check for leading / trailing whitespace. However, if we validate them as GUIDs we will also be able to prevent copying / pasting a GUID and making a mistake. In practice, even if the current GUIDs turns out to be an implementation detail, it's one that's not very likely to change. So I would vote for keeping the GUID check.

@rfranzke Any comments?

Reply:
@ialidzhikov Thanks! I added this reference to the code. I also reworked the validation to use a regex match for the above pattern, instead of `github.com/google/uuid` since I noticed that it allows uuids with formats different from the above that would likely fail on Azure (I checked one of them and it failed).

Reply:
> @ialidzhikov Thanks! I added this reference to the code. I also reworked the validation to use a regex match for the above pattern, instead of github.com/google/uuid since I noticed that it allows uuids with formats different from the above that would likely fail on Azure (I checked one of them and it failed).

Above I only shared the regex for the subsciption ID. Does the same regex apply for tenant ID and client ID?

Reply:
I feel a bit conflicted now. I checked our landscapes to see examples for `clientID` and `tenantID` and although they match the regex here - so no immediate issues from merging the PR - @ialidzhikov has a point we don't have official documentation for these regexes `cID` and `tID`. Not one that I could find at least.

Reply:
I discussed this with our Azure folks and since all 3 fields are GUIDs, the same restrictions apply equally. So the PR should be fine as it is.